2-Draft-Accepted/RFCxxxx-DSC-for-PowerShell7.md

```diff
@@ -0,0 +1,323 @@
+  ResourceID = "[File]DirectoryCopy";
+  ModuleVersion = "1.0";
+  ConfigurationName = "FileResourceDemo";
+};
+
+instance of MSFT_LogResource as $MSFT_LogResource1ref
+{
+  SourceInfo = "C:\\Users\\slee\\test\\file.dsc.ps1::15::9::Log";
+  ModuleName = "PsDesiredStateConfiguration";
+  Message = "Finished running the file resource with ID DirectoryCopy";
+  ResourceID = "[Log]AfterDirectoryCopy";
+  ModuleVersion = "1.0";
+  DependsOn = {
+    "[File]DirectoryCopy"
+  };
+  ConfigurationName = "FileResourceDemo";
+};
+
+instance of OMI_ConfigurationDocument
+{
+  Version="2.0.0";
+  MinimumCompatibleVersion = "1.0.0";
+  CompatibleVersionAdditionalProperties= {"Omi_BaseResource:ConfigurationName"};
+  Author="slee";
+  GenerationDate="11/05/2020 11:09:47";
+  GenerationHost="SLEE-DESKTOP";
+  Name="FileResourceDemo";
+};
+```
+
+Example JSON representation of generated configuration:
+
+```json
+{
+  "ConfigurationDocument": {
+    "GenerationDate": "11/05/2020 11:09:47",
+    "MinimumCompatibleVersion": "1.0.0",
+    "CompatibleVersionAdditionalProperties": [
+      "BaseResource:ConfigurationName"
+    ],
+    "Name": "FileResourceDemo",
+    "Version": "2.0.0",
+    "Author": "slee",
+    "GenerationHost": "SLEE-DESKTOP"
+  },
+  "LogResource": [
+    {
+      "SourceInfo": "C:\\\\Users\\\\slee\\\\test\\\\file.dsc.ps1::15::9::Log",
+      "ModuleVersion": "1.0",
+      "Message": "Finished running the file resource with ID DirectoryCopy",
+      "ResourceID": "[Log]AfterDirectoryCopy",
+      "ModuleName": "PsDesiredStateConfiguration",
+      "ConfigurationName": "FileResourceDemo",
+      "DependsOn": [
+        "[File]DirectoryCopy"
+      ]
+    }
+  ],
+  "FileDirectoryConfiguration": [
+    {
+      "ResourceID": "[File]DirectoryCopy",
+      "Ensure": "Present",
+      "Recurse": true,
+      "ModuleVersion": "1.0",
+      "DestinationPath": "C:\\\\Users\\\\Public\\\\Documents\\\\DSCDemo\\\\DemoDestination",
+      "ConfigurationName": "FileResourceDemo",
+      "SourceInfo": "C:\\\\Users\\\\slee\\\\test\\\\file.dsc.ps1::6::9::File",
+      "ModuleName": "PSDesiredStateConfiguration",
+      "Type": "Directory",
+      "SourcePath": "\\\\\\\\PullServer\\\\DemoSource"
+    }
+  ]
+}
+```
+
+Important differences:
+
+- CIM class name prefixes are removed as they are not needed
+- Instances are represented as an array of the type name
+- The comment at the start is dropped as that information is available within `ConfigurationDocument`
+
+### No longer require MOF schema by switching to improved class-based resources
+
+Although DSC resources can be written as WMI/OMI providers, this requires writing
+C/C++ code and compiling for specific Linux distros and processor architectures.
+Writing PowerShell script DSC resources currently still requires a matching schema
+file making it more complex. Instead, the current plan is to
+[only support PowerShell class-based DSC resources](https://github.com/PowerShell/PowerShell/issues/13731)
```

Comment:
Why is there no C# class-based DSC resource?

Reply:
It's certainly something that can be supported later if there is a need for it.  Initially, we're targeting making it simpler and PowerShell class-based DSC resource development model already exists while we'd have to create one for C#.

Reply:
Please correct me if I'm wrong but there really shouldn't be much additional work required supporting PowerShell and C# class-based resources. If the PowerShell class is being imported as a dynamic assembly, DSC can use the same logic to identify C# and PowerShell DSC resources.

Reply:
You are right, *theoretically* there really shouldn't be much additional work to support C# resources, however things get complex when we get into details because the existing code is rather difficult for historical reasons.
For example, for resource enumeration (`Get-DscResource`) current code does Not actually import PowerShell class as a dynamic assembly. It sends `psm1` with class to PS parser, retrieves AST and then traverses it looking for DSC attributes (e.g. `DscResourceAttribute`) and constructing internal resource metadata descriptions that are later (and Not in a straightforward way) returned as `Get-DscResource` results. The class is actually imported only at `Invoke-DscResource` stage.

Reply:
If it is difficult to add enhancements to the existing code base, then maybe its time to look at redesigning/refactoring while these major DSC changes are being proposed. As for `Get-DscResource` not importing the C# assembly until `Invoke-DscResource` is called, I believe that could be solved with a `MetadataLoadContext` as described in https://github.com/PowerShell/PowerShell/issues/6653.

Reply:
That is good info about `MetadataLoadContext` ; thank you.